CONTRIBUTING.md

```diff
@@ -0,0 +1,619 @@
+filePattern: "\\.(js|ts)x?$"      # .js, .jsx, .ts, .tsx files
+filePattern: "\\.(css|scss)$"     # .css and .scss files
+filePattern: "\\.ya?ml$"          # .yaml and .yml files
+```
+
+#### nodejs Provider (TypeScript/JavaScript Semantic Analysis)
+
+**Use for:**
+- Function/class/variable references
+- Import statements
+- JSX component usage
+- Semantic code analysis
+
+**Cannot find:**
+- Class methods (use builtin)
+- Object properties (use builtin)
+- Type annotations (use builtin)
+- JSX props (use builtin)
+
+**Example:**
+```yaml
+when:
+  nodejs.referenced:
+    pattern: "OldComponent"  # Finds actual symbol references
+```
+
+**Important:** nodejs provider does NOT support `filePattern`. It automatically searches all TypeScript/JavaScript files.
+
+#### java Provider (Java Semantic Analysis)
+
+**Use for:**
+- Java class/method/field references
+- Annotations
+- Import statements
+- Package declarations
+
+**Example:**
+```yaml
+when:
+  java.referenced:
+    pattern: "org.example.OldClass"
+```
+
+#### Other Providers
+
+- **go.referenced** - Go symbol references
+- **python.referenced** - Python symbol references
+- **builtin.xml** - XML element/attribute matching
+- **builtin.json** - JSON key/value matching
+- **builtin.hasTags** - Check for specific tags
+
+### Combining Providers for Complete Coverage
+
+**Best Practice:** Use semantic providers (nodejs, java, go, python) for symbol references, and builtin provider for patterns they cannot detect.
+
+**Example: Detecting React Component Migration**
+
+```yaml
+# nodejs provider - finds component imports and usage
+- ruleID: old-button-component-00000
+  when:
+    nodejs.referenced:
+      pattern: "OldButton"
+  message: |
+    OldButton component is deprecated.
+    Replace with NewButton.
+
+# builtin provider - finds prop usage (nodejs can't detect this)
+- ruleID: old-button-variant-prop-00001
+  when:
+    builtin.filecontent:
+      pattern: '<OldButton\s+variant="danger"'
+      filePattern: "\\.tsx?$"
+  message: |
+    The "danger" variant has been renamed to "destructive".
+```
+
+## Adding New Language Support
+
+### Steps to Add a New Provider
+
+1. **Create Provider Binary** - Implement LSP wrapper for the language server
+2. **Add Dockerfile** - Create container image for the provider
+3. **Update Makefile** - Add build targets
+4. **Add Example Project** - Create test project in `examples/`
+5. **Add Test Rules** - Create rules in `rule-example.yaml`
+6. **Update Demo Output** - Regenerate `demo-output.yaml` with all providers
+7. **Documentation** - Update README and this guide
+
+### Example: Adding File Extension Support
+
+When adding support for new file extensions to an existing provider (e.g., adding `.tsx` to the Node.js provider):
+
+**Typical files to modify:**
+1. **Provider configuration** - Add file extension to provider's supported types
+   - Example: `external-providers/generic-external-provider/pkg/server_config/providers.go`
```

Comment:
Do we want folks adding new providers using the generic-external-provider and committing to the tree? That may make sense for now, but I did want to call that out.